packages/devtools-source-map/src/scopes.js

```diff
@@ -0,0 +1,193 @@
+  ) {
+    return null;
+  }
+  return originalPosition.name;
+}
+
+// The `scope` has bindings that contains positions/references to the
+// generated names (these may has different original name). Going over all
+// these names and positions, to check what original name is listed there.
+// To make things more interesting the same generated name can potentially
+// be used for different original names, so building trackedBindings for all
+// names and positions.
+// Example of MappedScopeBindings:
+// ```
+// { type: 'block',
+//   bindings: {
+//     'current': 't',
+//     'start': 't',
+//     'end': 'e'
+//   } }
+// ```
+
+type TrackedBindings = { [originalName: string]:
+  {
+    generatedName: string,
+    bestPosition: SourceMapPosition
+  }
+};
+
+function applyBestNamePosition(
+  trackedBindings: TrackedBindings,
+  generatedName: string,
+  nameLocation: Location,
+  map: SourceMapConsumer,
+  currentPosition: SourceMapPosition
+): TrackedBindings {
+  const namePosition: SourceMapPosition = {
+    line: nameLocation.line,
+    column: nameLocation.column || 0
+  }
+  const originalName = getOriginalPositionName(map, namePosition);
+  if (!originalName) {
+    return trackedBindings;
+  }
+
+  let trackedBinding = trackedBindings[originalName];
+  if (!trackedBinding) {
+    // No best position was found for original name yet -- using first
+    // found position.
+    trackedBindings[originalName] = {
+      generatedName,
+      bestPosition: namePosition
+    };
+  } else if (comparePositions(namePosition, currentPosition) <= 0) {
+    // The position is before location, the more recient will have needed
+    // name assignment, also discard all positions past currentPosition.
+    if (
+      comparePositions(trackedBinding.bestPosition, namePosition) < 0 ||
+      comparePositions(trackedBinding.bestPosition, currentPosition) > 0
+    ) {
+      trackedBinding.generatedName = generatedName;
+      trackedBinding.bestPosition = namePosition;
+    }
+  } else {
+    // The position is past location, need to select more recient after
+    // specified location, but only if "before" bestPosition is not found.
+    if (
+      comparePositions(trackedBinding.bestPosition, namePosition) > 0 &&
+      comparePositions(trackedBinding.bestPosition, currentPosition) < 0
+    ) {
+      trackedBinding.generatedName = generatedName;
+      trackedBinding.bestPosition = namePosition;
+    }
+  }
+  return trackedBindings;
+}
+
+function applyTrackedBindings(
+  trackedBindings: TrackedBindings,
+  generatedName: string,
+  locations: Location[],
+  map: SourceMapConsumer,
+  currentPosition: SourceMapPosition
+): TrackedBindings {
+  return locations.reduce(
+    (trackedBindings, nameLocation) =>
+      applyBestNamePosition(
+        trackedBindings,
+        generatedName,
+        nameLocation,
+        map,
+        currentPosition
+      ),
+    trackedBindings
+  );
+}
+
+async function getLocationScopes(
```

Comment:
I'd like jsdoc for all new exported functions.  Lack of documentation is one of the bigger problems in this entire module already.